File: day3/3.hs
{-# OPTIONS_GHC -Wno-incomplete-patterns #-}
module Main where
import Data.Char (ord, isLower)
import Data.Functor ( (<&>) )
import Data.Set as S hiding (drop, take, map)

priority :: Char -> Int
priority x
    | isLower x = ord x - 96
    | otherwise = ord x - 38

readInput :: IO [[Int]]
readInput = readFile "./inputs/3" <&> lines <&> map (map priority)

solve1 :: [[Int]] -> Int
solve1 = sum . concatMap (\x -> S.toList (S.fromList (take (mid x) x) `S.intersection` S.fromList (drop (mid x) x)))
    where mid :: [Int] -> Int
          mid x = length x `div` 2

solve2 :: [[Int]] -> Int
solve2 [] = 0
solve2 (x:y:z:xs) = sum (S.toList  (S.fromList x `S.intersection` S.fromList y `S.intersection` S.fromList z)) + solve2 xs

main :: IO ()
main = do
    readInput >>= print . solve1
    readInput >>= print . solve2

File: 2022/day3/3.py
import os 
from collections import Counter

def priority(c):
    if c.islower():
        return ord(c) - 96 
    else:
        return ord(c) - 38

def solve1():
    total = 0 
    with open(os.getcwd() + "/day3/inputs/3") as f: 
        for line in f:
            line = line.strip()
            if line:
                x = list(map(priority, line))
                mid = len(x) // 2
                c1, c2 = Counter(x[0:mid]), Counter(x[mid:])
                for c in c1:
                    if c in c2:
                        total += c
    return total
def solve2():
    total = 0 
    with open(os.getcwd() + "/day3/inputs/3") as f: 
        lines = f.read().splitlines()
        i = 0 
        while i < len(lines):
            c1 = Counter(list(map(priority, lines[i])))
            c2 = Counter(list(map(priority, lines[i+1])))
            c3 = Counter(list(map(priority, lines[i+2])))
            for c in c1:
                if c in c2 and c in c3:
                    total += c
            i += 3         
    return total
print(solve1())
print(solve2())